test/hotspot/jtreg/runtime/cds/appcds/jvmti/CFLH/MultiReleaseJars.java

```diff
@@ -0,0 +1,177 @@
+            "class Bar {",
+            "    static {",
+            "        System.out.println(\"Hello from Bar\");",
+            "    }",
+            "}",
+        };
+        return sts;
+    }
+
+    static void writeFile(File file, String... contents) throws Exception {
+        if (contents == null) {
+            throw new java.lang.RuntimeException("No input for writing to file" + file);
+        }
+        try (
+             FileOutputStream fos = new FileOutputStream(file);
+             PrintStream ps = new PrintStream(fos)
+        ) {
+            for (String str : contents) {
+                ps.println(str);
+            }
+        }
+    }
+
+    /* version.jar entries and files:
+     * META-INF/
+     * META-INF/MANIFEST.MF
+     * Bar.class
+     * Main.class
+     * META-INF/versions/9/
+     * META-INF/versions/9/Bar.class
+     * META-INF/versions/9/Foo.class
+     */
+    static void createClassFilesAndJar() throws Exception {
+        String tempDir = CDSTestUtils.getOutputDir();
+        File baseDir = new File(tempDir + File.separator + "base");
+        File vDir    = new File(tempDir + File.separator + BASE_VERSION_STRING);
+
+        baseDir.mkdirs();
+        vDir.mkdirs();
+
+        File fileFoo = TestCommon.getOutputSourceFile("Foo.java");
+        writeFile(fileFoo, getFoo());
+        JarBuilder.compile(vDir.getAbsolutePath(), fileFoo.getAbsolutePath(), "--release", BASE_VERSION_STRING);
+
+        File fileMain = TestCommon.getOutputSourceFile("Main.java");
+        writeFile(fileMain, getMain());
+        JarBuilder.compile(baseDir.getAbsolutePath(), fileMain.getAbsolutePath());
+        File fileBar = TestCommon.getOutputSourceFile("Bar.java");
+        writeFile(fileBar, getBar());
+        JarBuilder.compile(baseDir.getAbsolutePath(), fileBar.getAbsolutePath());
+        JarBuilder.compile(vDir.getAbsolutePath(), fileBar.getAbsolutePath(), "--release", BASE_VERSION_STRING);
+
+        String[] meta = {
+            "Multi-Release: true",
+            "Main-Class: Main"
+        };
+        File metainf = new File(tempDir, "mf.txt");
+        writeFile(metainf, meta);
+
+        JarBuilder.build("multi-version", baseDir, metainf.getAbsolutePath(),
+            "--release", BASE_VERSION_STRING, "-C", vDir.getAbsolutePath(), ".");
+
+    }
+
+    static void checkExecOutput(OutputAnalyzer output, String expectedOutput) throws Exception {
+        try {
+            TestCommon.checkExec(output, expectedOutput);
+        } catch (java.lang.RuntimeException re) {
+            String cause = re.getMessage();
+            if (!expectedOutput.equals(cause)) {
+                throw re;
+            }
+        }
+    }
+
+    public static void main(String... args) throws Exception {
+        // create multi-version.jar which contains Main.class, Foo.class and Bar.class.
+        // Foo.class has only one version: base version 9.
+        // Bar.class has two versions: base version 9 and default version.
```

Comment:
Since there's no default version for Foo, does the class loader try to get the highest version (9 in this case) that's the same or below the current  JDK's version? I think such a comment in the test case would be helpful.

Also, perhaps the test should look like this:

```
     META-INF/versions/9/Foo.class
     META-INF/versions/11/Foo.class
```

where the different Foo classes print a different string. We should assert that we get version 11 instead of 9.

Reply:
I've updated the test. Instead of version 11, I've used the major JDK version.